### PR TITLE
Add mysql_query module

### PR DESCRIPTION
**Pull request by nerzhul**
This module permit to run single query or file to a remote MySQL server.
All syntax is included in the PR :)

### CODE REVIEW

**Comment by sivel** [comment]
This pull request has failed Travis checks due to an incorrect `version_added`.  This should be updated to reflect `2.1`

**Comment by nerzhul** [comment]
Fixed

**Comment by nerzhul** [comment]
@sivel fixed :)

**Comment by tersmitten** [comment]
See #980

**Comment by tersmitten** [comment]
Great addition!

**Comment by nerzhul** [comment]
Thanks :) I hope this will be approved :) i use it in production since today to replace my old shell: mysql < {{ file }} or shell: mysql -e 'FLUSH TABLES' ...

**Comment by gregdek** [comment]
Thanks @nerzhul for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

**Comment by nerzhul** [comment]
@gregdek nobody wants to review it ? :)

**Comment by tersmitten** [comment]
I'll try to find some time to review it

**Comment by tersmitten** [comment]
@Jmainguy Can you also have a look at this?

File: database/mysql/mysql_query.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+"""
+
+Ansible module to manage mysql variables
```
Comment (Jmainguy) [inline database/mysql/mysql_query.py]: This module is actually to execute query's right?

**Comment by Jmainguy** [comment]
This is a great module to add, code looks good from a quick review. I will run some tests later today for good measure and add a 'shipit' then, nice work.

File: database/mysql/mysql_query.py (hunk 1 above)
Comment (nerzhul) [inline database/mysql/mysql_query.py]: You note that i copy from another mysql module, i fix it now :)

**Comment by nerzhul** [comment]
@Jmainguy fixed !

**Comment by Jmainguy** [comment]
@nerzhul I tested it, and it works very well. (Tested on centos7 with mariadb).

A feature request I would love to see, would be to add --diff support to show the result of the query (for instance if a select was used), and storing that variable so that a separate task could then make use of that value.

Great work man. @gregdek 'shipit'

**Comment by nerzhul** [comment]
Thanks @Jmainguy , this addition could be good, but maybe in a second time, i think :)

Another voter present ? :)

**Comment by tersmitten** [comment]
Somehow I have trouble with INSERT queries:

```
---
- hosts: foo.dev
  tasks:
    - name: mysql query
      mysql_query:
        db: test
        query: >
          CREATE TABLE IF NOT EXISTS `foo` (
            `id` INT(6) UNSIGNED AUTO_INCREMENT PRIMARY KEY,
            `firstname` VARCHAR(30) NOT NULL,
            `lastname` VARCHAR(30) NOT NULL,
            `email` VARCHAR(50),
            `reg_date` TIMESTAMP
          );
      ignore_errors: true

    - name: mysql query
      mysql_query:
        db: test
        query: >
          INSERT INTO `foo` (`id`, `firstname`, `lastname`, `email`, `reg_date`) VALUES (NULL, 'foo', 'bar', 'example.com', NOW());

    - name: mysql query
      mysql_query:
        db: mysql
        query: >
          FLUSH TABLES;

    - name: mysql query file
      mysql_query:
        db: test
        query_file: /tmp/foo.sql
      ignore_errors: true
```

```
# cat /tmp/foo.sql
CREATE TABLE IF NOT EXISTS `bar` LIKE `foo`;
INSERT INTO `bar` (`id`, `firstname`, `lastname`, `email`, `reg_date`) VALUES (NULL, 'foo', 'bar', 'example.com', NOW());
```

Create table statements seem to work, but tables `foo` and `bar` remain empty. Is it possible that a `.commit` is missing?

Also I have trouble with queries that generate a warning (e.g. `CREATE TABLE IF NOT EXISTS foo`). Is that the expected behavior?

```
PLAY ***************************************************************************

TASK [setup] *******************************************************************
ok: [foo.dev]

TASK [mysql query] *************************************************************
fatal: [foo.dev]: FAILED! => {"changed": false, "failed": true, "msg": "Table 'foo' already exists"}
...ignoring

TASK [mysql query] *************************************************************
changed: [foo.dev]

TASK [mysql query] *************************************************************
ok: [foo.dev]

TASK [mysql query file] ********************************************************
fatal: [foo.dev]: FAILED! => {"changed": false, "failed": true, "msg": "(1065, 'Query was empty')"}
...ignoring

PLAY RECAP *********************************************************************
foo.dev               : ok=5    changed=1    unreachable=0    failed=0   
```

**Comment by nerzhul** [comment]
@tersmitten it's the standard mysqlDB behaviour, not module behaviour. I will look at your tests to see what could be wrong there to improve error handling

**Comment by tersmitten** [comment]
I know, but in order to let the INSERT (UPDATE et cetera)'s succeed you need to commit the cursor.

In order to ignore warnings you can do this:

```
warnings.filterwarnings('ignore', category=MySQLdb.Warning)
```

[See](https://github.com/nadley/ansible-modules-extras/commit/290254d607268439c573722cd4cae7fe8a985b53)

**Comment by nerzhul** [comment]
@tersmitten you are right, sorry i was tired yesterday :)

I fixed the warning issue

**Comment by nerzhul** [comment]
@tersmitten all issues are now fixed ! thanks for your tests

**Comment by nerzhul** [comment]
And to finish i tested all your playbook and it works like a charm

**Comment by tersmitten** [comment]
Indeed, nice work.

@gregdek 'shipit'

**Comment by nerzhul** [comment]
@gregdek @tersmitten when a module has two shitip, when will it be ship into the main repository ? Before a release ?

**Comment by Jmainguy** [comment]
@gregdek @robynbergeron this is ready for inclusion/core review, can one of you update this.

**Comment by nerzhul** [comment]
Don't forget to tag for inclusion !

**Comment by gregdek** [comment]
Thanks @nerzhul and reviewers. Putting into shipit for final review.

**Comment by nerzhul** [comment]
Thanks @gregdek

**Comment by abadger** [comment]
We discussed this at today's meeting and it turned into a discussion of the very fundamental question of where to draw the line between creating a module versus just using the command or shell module.  We'll have to continue this discussion at next week's meeting as there were a plethora of thoughts on the matter but no consensus.  Some of the view points would favor this module going in and other viewpoints would see this module excluded from the ansible tarball (an open question is how we would like to see it distributed if that is the case).  Some of the ideas that were floated today (note -- some of these ideas were popular, others were unpopular.  Many of them are mutually exclusive.):
- Modules should only be written if the module can be idempotent or support check-mode (neither of which are well supported by the command module)
- Modules are not for returning information
- Modules are for simplifying the user experience.
- If it can be done in the shell module then it shouldn't be its own module
- Modules should not be thin wrappers of cli tools but can wrap more complex interactions
- Users should be able to clearly tell which modules are idempotent and which aren't -- Idempotence is desired for playbooks but non-idempotent modules are still good for new users using /usr/bin/ansible to simplify common ad hoc tasks.
- There needs to be a place for modules which would be kept out by these rules
- The extras repository is a place for modules which do not fit these rules (we should apply these rules to core but we need extras to be a place where people can experiment and come up with things that are needed even though they don't fit the current rules).
- Having a module that can shoot you in the foot (drop tables, insert garbage) is not good.
- Many modules allow you to shoot yourself in the foot.
- Having specific modules for each database flavour instead of a generic one that can apply to many sql databases is the wrong design.
- Query functionality should be a lookup plugin or a facts module.

This is kinda a collective brain dump.  It shows that we have a lot of discussion to do in the next meeting to reconcile the major themes here.

**Comment by nerzhul** [comment]
@abadger thanks for your big reply.
This module is not for querying but for executing. It permit to have a scenario with ansible well played commands when you are putting things to production. The problem with shell is it's a generic crap wrapper for non handled ansible things. Shell permit to do everything, but users may also want (and it's some enterprise goal) drop all those shell craps to have a common ansible interface with modules.

**Comment by thaumos** [comment]
@nerzhul, when you say executing, are you inferring INSERT only, or SELECT as well?  I need clarity for myself. 

Thanks for responding back on this!  We really appreciate your continued involvement in this!

**Comment by nerzhul** [comment]
@thaumos the objective of this module is to do modifications on database, not querying, this is not the purpose of this module.
This module is for orchestration & deployment not for idem potence. If you want idem potence into BDD stats, i recommend writing a mysql_record module to maintain a specific record into DB

**Comment by thaumos** [comment]
@nerzhul then it could be argued the name of the module is misleading... If you intend to only do INSERTS that is.  In my mind, this module enables a user to do a mysql_query (INSERT or SELECT). 😄 

I suggest, and we would love, if you joined our meeting next week to discuss this with us!  

When @abadger mentioned the list of things, as he said, it is not all inclusive, but one of the things mentioned was "idempotency."  Knowing databases, there isn't a way to make any action idempotent.  So that point does and does not apply.  Like he said, it was a collective brain dump.

**Comment by nerzhul** [comment]
@thaumos the problem is INSERT, SELECT, UPDATE, DELETE and more are all queries, it's the database noun it's why i choose it. 
This module is not for doing select-only queries, it's for database modifications.

@abadger as you said check mode is missing, i will read the doc and add it, it could be easy to implement because of MySQL transactions

**Comment by Jmainguy** [comment]
@nerzhul I am thinking check mode would basically just be testing the connection to mysql, and then returning changed if it can connect, but I will leave that for you to decide of course.

@abadger @thaumos I can appreciate the discussion being had, In that big list of info dump, I would say we have many modules that are currently merged that do not follow those conventions. I think ansible deciding on what modules should and should not be is a good idea, however once we make those rules we would have to go back and boot a bunch of modules out, hopefully I can make the next meeting as this discussion is very interesting.

**Comment by bcoca** [comment]
@Jmainguy  there are many different opinions reflected in that dump. 

That existing modules violate newer rulesmostly means that the experience of having those modules are the reason for theses rules. They are in place to stop the problem from growing.
Instead of booting them I prefer to update them to reflect the newer rules, as we have been doing in many cases already.  Some modules cannot be updated so we are trying to create new ones that do follow the rules and provide the same functionality in views of deprecating/removing the offenders.

As for the decision on which modules go in or not, I would argue that those that assume the support burden should have a bigger say. As the expectations are that we support anything shipped, even though that might be unreasonable, no matter how much we point at extras being more community driven.

I'm all for giving a lot of latitude and leeway for extras, but there are things I think we should not ship ourselves and I'm fine with others publishing in galaxy/github/anywhere else where it is clearly and unequivocally unrelated to our code base. If we ever push 'extras' to the point of not being shipped or packaged by Ansible, I'm happy to not give any restrictions.

**Comment by Jmainguy** [comment]
@bcoca Understood, and yeah I definitely think anything shipped under the ansible name is supported by ansible/redhat including extras. There is precedent for non officially supported modules out there, https://github.com/cmprescott/ansible-xml being one I am aware of. 

I think ansible defining these rules on modules will really help with future development.

**Comment by bcoca** [comment]
most are already official http://docs.ansible.com/ansible/developing_modules.html, we are still 'fine tunnig' some of the newer ones.

**Comment by nerzhul** [comment]
@Jmainguy @abadger i added check mode to the module, whichonly connects to database without performing requests and return changed=True. Is this good for you ?

**Comment by tima** [comment]
@nerzhul I would expect the check mode on a module like this to check if an INSERT/UPDATE/DELETE is needed (do these rows exists? are those column values different from the ones I want update?) and report that status without performing that actual operation.

**Comment by nerzhul** [comment]
@tima yes, i only refresh the PR to show it's active :). It's the goal, but i didn't know if it could be possible or not, i need to do more tests

**Comment by nerzhul** [comment]
@tima @Jmainguy i improved the module this morning, as i know feature freeze happened.
- It don't log password anymore
- check mode returns rows affected by executing queries but not commiting them
- query_file ignore empty lines and comment lines

**Comment by nerzhul** [comment]
Then another review is needed and @abadger i hope you will prefer this version :)

**Comment by nerzhul** [comment]
I added a little fix provided by @Farzy to fix Dos lines and comments starting with whitespaces

**Comment by gregdek** [comment]
Thanks @nerzhul for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by nerzhul** [comment]
ready_for_review @gregdek

**Comment by gregdek** [comment]
Thanks @nerzhul for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by tima** [comment]
gun_meet_foot

This module still does not support idempotency and makes doing tremendous harm to a database way too easy. It does little more than calling command/shell. Shipping this module with Ansible communicates an endorsement I don't think should be made. Like @bcoca  I'm fine with whatever being published anywhere else where it is clearly and unequivocally unrelated to the code base here.

**Comment by Jmainguy** [comment]
There are many modules currently in Core/Extras that do not support idempotency, and allow destruction of the system if used in a certain way.

I am in favor of this module.

**Comment by tima** [comment]
@Jmainguy: Two wrongs don't make a right.

**Comment by nerzhul** [comment]
@tima this module is for doing sql queries in ansible way, idempotency is not possible every time, but it's useful because in a playbook you could have a need to do, for example, a delete on a table everytime your play it. This delete guarentee idempotency over the result, you will always delete all results matching the query.

**Comment by tima** [comment]
I want to clarify my previous comments. 

I understand the desire to automate database table operations and I'm not objecting to or discounting the desire for that functionality. My biggest issue is that this module is so wide open to doing anything/everything you throw at it with no safeties or other logic. Idempotentcy helps to an extent, but is not the only way to that end. It's still a really good idea and I highly encourage it, but its not _the_ issue.

So I understand the desire to drop (delete) a table as part of a playbook run. I am not objecting to that need  -- it's that there should be a module that specifically handles table add/drops and only that, not any SQL statement one sends it.

**Comment by nerzhul** [comment]
@tima i agree with idempotency and i agree that should be used everytime it's possible, but for this goal i don't think it's possible, because a database is not really an idempotent system.

For a single table yes we could have a thing, for a single record it's possible too, but for a SQL incoming from external source or which should be played at each playbook run and doesn't give exactly same situation on two runs (for example SQL with updated datas to run on a table every day)

**Comment by ryansb** [comment]
Strictly a database can be used in an idempotent way, but it requires a certain data model that I don't know if we can guarantee. For example:

```
INSERT INTO `foo` (`id`, `firstname`, `lastname`, `email`, `reg_date`) 
    VALUES (NULL, 'foo', 'bar', 'example.com', NOW())
    ON DUPLICATE KEY UPDATE id=id;
```

That will insert the record, unless the `id` column conflicts in which case it will update `id=id` which will be optimized out by the MySQL query engine. Of course:
1. This only works if `id` has a primary key constraint
2. And ON DUPLICATE is supported
3. And the user's data model makes sense for this use case

**Pull request closed without merging**